gulpfile.js

```diff
@@ -297,6 +303,216 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
+            done();
+        }
+    });
+
+});
+
+// Deploy cesium to s3
+function deployCesium(bucketName, uploadDirectory, cacheControl) {
+    var readFile = Promise.promisify(fs.readFile);
+    var gzip = Promise.promisify(zlib.gzip);
+    var concurrencyLimit = 2000;
+
+    var s3 = new Promise.promisifyAll(new aws.S3({
+        maxRetries : 10,
+        retryDelayOptions : {
+            base : 500
+        }
+    }));
+
+    var existingBlobs = [];
+    var totalFiles = 0;
+    var uploaded = 0;
+    var skipped = 0;
+
+    return listAll(s3, bucketName, uploadDirectory, existingBlobs)
+    .then(function() {
+        return globby([
+            'Apps/**',
+            'Build/**',
+            'Source/**',
+            'Specs/**',
+            'ThirdParty/**',
+            '*.md',
+            'favicon.ico',
+            'gulpfile.js',
+            'index.html',
+            'logo.png',
+            'package.json',
+            'server.js',
+            'web.config'
+        ], {
+            dot : true, // include hidden files
+            nodir : true // only directory files
+        });
+    }).then(function(files) {
+        return Promise.map(files, function(file) {
+            var blobName = uploadDirectory + '/' + file;
+            var mimeLookup = getMimeType(blobName);
+            var contentType = mimeLookup.type;
+            var compress = mimeLookup.compress;
+            var contentEncoding = compress ? 'gzip' : undefined;
+            var etag;
+
+            totalFiles++;
+
+            return readFile(file)
+            .then(function(content) {
+                return compress ? gzip(content) : content;
+            })
+            .then(function(content) {
+                // compute hash and etag
+                var hash = crypto.createHash('md5').update(content).digest('hex');
+                etag = crypto.createHash('md5').update(content).digest('base64');
+
+                var index = existingBlobs.indexOf(blobName);
+                if (index <= -1) {
+                    return content;
+                }
+
+                // remove files as we find them on disk
+                existingBlobs.splice(index, 1);
+
+                // get file info
+                return s3.headObjectAsync({
+                    Bucket : bucketName,
+                    Key : blobName
+                })
+                .then(function(data) {
+                    if (data.ETag !== ('"' + hash + '"') ||
+                        data.CacheControl !== cacheControl ||
+                        data.ContentType !== contentType ||
+                        data.ContentEncoding !== contentEncoding) {
+                        return content;
+                    }
+
+                    // We don't need to upload this file again
+                    skipped++;
+                    return undefined;
+                })
+                .catch(function() {
+                    return content;
```

Comment:
I'm having a hard time figuring out how this catch block ever gets triggered.  The only case I can think of is if a file is in `existingBlobs` but is deleted by something else before we get to the `headObjectAsync` call.  Is there another case I'm not thinking of?  Part of me things we should just take it out since it might mask another error.  Why is it needed?